pkg/storage/s3/s3_test.go

```diff
@@ -100,16 +105,19 @@ func TestGetConfigCustomRegionEndpoint(t *testing.T) {
 	})
 	listers := testBuilder.BuildListers()
 
-	config, err := GetConfig(listers)
+	s3Driver := &driver{
+		Listers: listers,
+		Config:  &imageregistryv1.ImageRegistryConfigStorageS3{},
+	}
+	config, err := s3Driver.UpdateEffectiveConfig()
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	expected := &S3{
-		AccessKey:      "access",
-		SecretKey:      "secret",
-		Region:         "example",
-		RegionEndpoint: "https://s3.example.com",
+	expected := &imageregistryv1.ImageRegistryConfigStorageS3{
+		Region:             "example",
+		RegionEndpoint:     "https://s3.example.com",
+		VirtualHostedStyle: true,
```

Comment:
Was this test wrong before? I mean, why do we expect `VirtualHostedStyle` to be true now?

Reply:
I hope I can explain this :)

The current test in the master branch calls and tests s3.GetConfig(), but the actual s3 package really uses s3.updateConfigAndGetCredentials() (which in turn does call s3.GetConfig()). And in s3.updateConfigAndGetCredentials() there is an extra step where the VirtualHostedStyle is set on the config depending on various values. https://github.com/openshift/cluster-image-registry-operator/blob/master/pkg/storage/s3/s3.go#L124-L129

The restructured code now has the single s3.UpdateEffectiveConfig() call to do the same work that s3.updateConfigAndGetCredentials() used to do. And the new s3.UpdateEffectiveConfig() now does the setting of the VirtualHostedStyle that only s3.updateConfigAndGetCredentials() used to do.

So now the test case in this PR is updated to call/test s3.UpdateEffectiveConfig() and the VirtualHostedStyle value is being set. And that is why I had to update the test case.

I honestly don't know what this value of VirtualHostedStyle accomplishes, and I just tried to preserve the existing behavior. I spent a lot of time trying to figure out why it is necessary to set VirtualHostedStyle, but I don't believe I was able to fully understand it.

As I re-read the code just now to write up this reply, I see that the current s3.updateConfigAndGetCredentials() actually tests the condition for both Config.Region and Config.RegionEndpoint being unset before it goes down the code path to potentially set Config.VirtualHostedStyle, but my restructuring only checks against Region being unset. I *think* it doesn't affect the way things work as this code really only ever runs the first time we are populating the config, but I could change it to more faithfully reproduce what the current code does.

WDYT?

Reply:
To be honest I also don't fully understand the impacts of VirtualHostedStyle and when we should leverage it and when we don't (it seems like the user can activate it manually as well). I know it has something to do with the way we reach the bucket: with VirtualHostedStyle the bucket is accessible through `<bucketname>.s3.amazon.com` instead of `s3.amazonaws.com/<bucketname>`. That being said I think we can agree that the least we change the current behavior the better, specially because @dmage, that would be able to explain us exactly what that means and in what circumstances we need to use it, is off. BTW, I think that the access without VirtualHostedStyle (for new buckets) has been deprecated: https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/

Reply:
:+1: i put back the condition to match the previous behavior so only if Region and RegionEndpoint were zero-length would it go down that codepath.